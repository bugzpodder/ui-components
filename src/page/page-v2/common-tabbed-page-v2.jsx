// @flow
import React, { type Node } from "react";
import { CommonPageV2 } from "./common-page-v2";
import { TabsComponent } from "./components/tabs-component";

type Props = {
  /**
   * The array of objects used to define the tabs displayed in the header, and their corresponding pages.
   * The `key` property is used to determine which tab is selected.
   */
  pageConfigs: Array<PageConfigV2>,
  /** Takes the handler used to switch between tabs, based on the tab's value */
  onChangeActiveTab: string => any,
  /** The value of the tab header. Used to define which tab is open */
  activeTab: string | false,
  /** Page title */
  title?: Node,
  /** Subtitle shown under the `title` */
  subtitle?: Node,
  /**
   *
   * Provides classNames to the page and its sub-components. Options include:
   *
   *  - `root`
   *
   *  - `header` (applied to entire header container which contains headerActions, title, subtitle, and special actions)
   *
   *  - `headerActions` (applied to the container around the header actions)
   *
   *  - `primaryActions`
   *
   *  - `secondaryActions`
   *
   *  - `tabs` (applied to the container around the header tabs)
   *
   *  - `tab` (applied to each individual tab)
   *
   *  - `titleContainer` (applied to the container around the title and subtitle)
   *
   *  - `title`
   *
   *  - `subtitle`
   *
   *  - `content` - (applied to the container around the content)
   */
  classes?: CommonTabbedPageV2Classes,
  /** Takes a `node` to show on the page */
  children?: Node,
  /** Displays a non-interactive loading animation */
  isLoading?: boolean,
  /** Primary actions to display on the header */
  primaryActions?: Array<HeaderAction>,
  /** Secondary actions to display in the secondary actions menu */
  secondaryActions?: Array<HeaderAction>,
  /** Props applied to the `Tabs` container */
  tabProps?: Object,
  /** Displays an non-interactive loading animation over the page content when true */
  isLoading?: boolean,
  /** When `false`, tab labels will be set to in-line. Defaults to true */
  wrapTabLabels?: boolean,
};

/**
 * `CommonTabbedPage` provides a component for a page with a flush card header,
 * a tab bar, and an optional side menu. Note: if you are using this component
 * in a new environment, we recommend building a new child component for your
 * environment; take a look at the `TabbedLimsPage` component as an example.
 */
export const CommonTabbedPageV2 = (props: Props) => {
  const {
    classes = {},
    activeTab,
    pageConfigs,
    onChangeActiveTab,
    children,
    isLoading = false,
    wrapTabLabels = true,
    tabProps = {},
    ...commonPageProps
  } = props;
  if (!isLoading && !pageConfigs) {
    throw new Error("pageConfigs is required");
  }
  const selectedTab = pageConfigs.find(tab => tab.key === activeTab);
  if (!isLoading && activeTab && !selectedTab) {
    console.error("no pageConfig keys match the provided activeTab");
  }
  let PageContent = null;
  if (selectedTab && selectedTab.Component) {
    const SelectedComponent = selectedTab.Component;
    PageContent = <SelectedComponent {...(selectedTab.componentProps || {})} />;
  }
  const { tabs, tab, ...commonPageClasses } = classes;
  return (
    <CommonPageV2
      {...commonPageProps}
      classes={{ ...commonPageClasses }}
      isLoading={isLoading}
      centerHeader={
        pageConfigs &&
        pageConfigs.length > 0 && (
          <TabsComponent
            wrapTabLabels={wrapTabLabels}
            activeTab={(!isLoading && activeTab) || false}
            onChangeActiveTab={onChangeActiveTab}
            classes={{ tabs, tab }}
            pageConfigs={pageConfigs}
            tabProps={tabProps}
          />
        )
      }
    >
      {PageContent}
      {children}
    </CommonPageV2>
  );
};
