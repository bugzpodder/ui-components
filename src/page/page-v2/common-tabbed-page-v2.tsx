import React, { ReactNode } from "react";
import { ClickableItem } from "../../types/dropdown";
import { CommonPageV2 } from "./common-page-v2";
import {
  CommonTabbedPageV2Classes,
  HeaderAction,
  PageConfigV2,
} from "../../types/card";
import { TabProps } from "@material-ui/core/Tab";
import { TabsComponent } from "./components/tabs-component";

type Props = {
  /**
   * The array of objects used to define the tabs displayed in the header, and their corresponding pages.
   * The `key` property is used to determine which tab is selected.
   */
  pageConfigs: PageConfigV2[];
  /** Takes the handler used to switch between tabs, based on the tab's value */
  onChangeActiveTab: (x0: string) => any;
  /** The value of the tab header. Used to define which tab is open */
  activeTab: string | false;
  /** Page title */
  title?: ReactNode;
  /** Subtitle shown under the `title` */
  subtitle?: ReactNode;
  /**
   *
   * Provides classNames to the page and its sub-components. Options include:
   *
   *  - `root`
   *
   *  - `header` (applied to entire header container which contains headerActions, title, subtitle, and special actions)
   *
   *  - `headerActions` (applied to the container around the header actions)
   *
   *  - `primaryActions`
   *
   *  - `secondaryActions`
   *
   *  - `tabs` (applied to the container around the header tabs)
   *
   *  - `tab` (applied to each individual tab)
   *
   *  - `titleContainer` (applied to the container around the title and subtitle)
   *
   *  - `title`
   *
   *  - `subtitle`
   *
   *  - `content` - (applied to the container around the content)
   */
  classes?: CommonTabbedPageV2Classes;
  /** Takes a `node` to show on the page */
  children?: ReactNode;
  /** Primary actions to display on the header */
  primaryActions?: HeaderAction[];
  /** Secondary actions to display in the secondary actions menu */
  secondaryActions?: ClickableItem[];
  /** Props applied to the `Tabs` container */
  tabProps?: Partial<TabProps>;
  /** Displays an non-interactive loading animation over the page content when true */
  isLoading?: boolean;
  /** When `false`, tab labels will be set to in-line. Defaults to true */
  wrapTabLabels?: boolean;
};

/**
 * `CommonTabbedPage` provides a component for a page with a flush card header,
 * a tab bar, and an optional side menu. Note: if you are using this component
 * in a new environment, we recommend building a new child component for your
 * environment; take a look at the `TabbedLimsPage` component as an example.
 */
export const CommonTabbedPageV2: React.FC<Props> = props => {
  const {
    classes = {},
    activeTab,
    pageConfigs,
    onChangeActiveTab,
    children,
    isLoading = false,
    wrapTabLabels = true,
    tabProps = {},
    ...commonPageProps
  } = props;
  if (!isLoading && !pageConfigs) {
    throw new Error("pageConfigs is required");
  }
  const selectedTab = pageConfigs.find(tab => tab.key === activeTab);
  if (!isLoading && activeTab && !selectedTab) {
    console.error("no pageConfig keys match the provided activeTab");
  }
  let PageContent: ReactNode = null;
  if (selectedTab && selectedTab.Component) {
    const SelectedComponent = selectedTab.Component;
    PageContent = <SelectedComponent {...(selectedTab.componentProps || {})} />;
  }
  const { tabs, tab, ...commonPageClasses } = classes;
  return (
    <CommonPageV2
      {...commonPageProps}
      classes={{ ...commonPageClasses }}
      isLoading={isLoading}
      centerHeader={
        pageConfigs &&
        pageConfigs.length > 0 && (
          <TabsComponent
            wrapTabLabels={wrapTabLabels}
            activeTab={(!isLoading && activeTab) || false}
            onChangeActiveTab={onChangeActiveTab}
            classes={{ tabs, tab }}
            pageConfigs={pageConfigs}
            tabProps={tabProps}
          />
        )
      }
    >
      {PageContent}
      {children}
    </CommonPageV2>
  );
};
