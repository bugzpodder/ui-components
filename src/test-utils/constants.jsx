// @flow
import React from "react";
import { EDC, PIPELINE } from "@grail/lib";

export const COUNTRIES: Array<Object> = [
  { label: "Afghanistan" },
  { label: "Aland Islands" },
  { label: "Albania" },
  { label: "Algeria" },
  { label: "American Samoa" },
  { label: "Andorra" },
  { label: "Angola" },
  { label: "Anguilla" },
  { label: "Antarctica" },
  { label: "Argentina" },
  { label: "Armenia" },
  { label: "Aruba" },
  { label: "Australia" },
  { label: "Austria" },
  { label: "Azerbaijan" },
  { label: "Bahamas" },
  { label: "Bahrain" },
  { label: "Bangladesh" },
  { label: "Barbados" },
  { label: "Belarus" },
  { label: "Belgium" },
  { label: "Belize" },
  { label: "Benin" },
  { label: "Bermuda" },
  { label: "Bhutan" },
].map(suggestion => ({
  label: suggestion.label,
  value: suggestion.label.toUpperCase().replace(" ", "_"),
}));

export const GRAIL = "https://grail.com/wp-content/uploads/2016/12/cropped-final-favicon-32x32.png";

export const EXAMPLE_TABLE_DATA = [
  { word: "Argle-bargle", origin: "Britain" },
  { word: "Dwaal", origin: "South Africa" },
  { word: "Floccinaucinihilipilification", origin: "General English" },
  { word: "Lollygag", origin: "North American" },
  { word: "Mouse Potato", origin: "General English" },
  { word: "Ogdoad", origin: "General English" },
  { word: "Sesquipedalian", origin: "General English" },
  { word: "Flibbertigibbet", origin: "Britain" },
  { word: "Curmudgeon", origin: "N/A" },
  { word: "Nudiustertian", origin: "Latin" },
  { word: "Tittynope", origin: "General English" },
  { word: "Pauciloquent", origin: "Britain" },
  { word: "Pneumonoultramicroscopicsilicovolcanoconiosis", origin: "General English" },
  { word: "Bijoux", origin: "Archaic" },
  { word: "Fartlek", origin: "Swedish" },
  { word: "Cruciverbalist", origin: "General English" },
  { word: "Erinaceous", origin: "General English" },
];

export const LARGE_EXAMPLE_TABLE_DATA = [];
for (let i = 0; i < 1000; i++) {
  // Create about 17,000 records.
  LARGE_EXAMPLE_TABLE_DATA.push(...EXAMPLE_TABLE_DATA);
}

export const EXTRA_LARGE_EXAMPLE_TABLE_DATA = [];
for (let i = 0; i < 100; i++) {
  // Create about 1,700,000 records
  EXTRA_LARGE_EXAMPLE_TABLE_DATA.push(...LARGE_EXAMPLE_TABLE_DATA);
}

export const ADAMS_DATA = [
  {
    date: "1848-02-23T16:20:00Z",
    content: <div>John Quincy Adams dies.</div>,
  },
  {
    date: "1818-04-10T12:00:00Z",
    content: (
      <div>
        John Quincy Adams tries to dig to the middle of the Earth to conduct trade with the people living there.
      </div>
    ),
  },
  {
    date: "1767-07-11T12:20:00Z",
    content: <div>John Quincy Adams is born.</div>,
  },
];

export const EMU_DATA = [
  {
    date: "1932-12-10T16:20:00Z",
    content: <div>War ends and is declared a victory for the emus.</div>,
  },
  {
    date: "1932-11-08T12:00:00Z",
    content: <div>Over 2500 rounds had been fired with the media claiming that only a few emus had been killed.</div>,
  },
  {
    date: "1932-11-02T12:20:00Z",
    content: <div>Sir George Pearce wages war on the local population of emus in Australia.</div>,
  },
];

export const ALPHABET_ANIMALS_DATA: Array<Array<?string>> = [
  ["alligator", "bat", "cat", "dog", "elephant", "fox"],
  ["goat", "horse", "iguana", "jaguar", "kangaroo", "lion"],
  ["monkey", "narwhal", "octopus", "penguin", "quail", "raccoon"],
  ["seal", "tiger", "umbrella bird", "vulture", "whale", "x-ray tetra"],
  ["yak", "zebra", null, null, null, null],
];

export const TEST_EXTERNAL_DOMAINS: Map<string, string> = new Map().set(EDC, "edc.com").set(PIPELINE, "pipeline.com");
