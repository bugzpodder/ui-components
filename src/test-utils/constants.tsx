import React from "react";

export const COUNTRY_NAMES = [
  "Afghanistan",
  "Akrotiri",
  "Albania",
  "Algeria",
  "American Samoa",
  "Andorra",
  "Angola",
  "Anguilla",
  "Antarctica",
  "Antigua and Barbuda",
  "Argentina",
  "Armenia",
  "Aruba",
  "Ashmore and Cartier Islands",
  "Australia",
  "Austria",
  "Azerbaijan",
  "Bahamas, The",
  "Bahrain",
  "Bangladesh",
  "Barbados",
  "Bassas da India",
  "Belarus",
  "Belgium",
  "Belize",
  "Benin",
  "Bermuda",
  "Bhutan",
  "Bolivia",
  "Bosnia and Herzegovina",
  "Botswana",
  "Bouvet Island",
  "Brazil",
  "British Indian Ocean Territory",
  "British Virgin Islands",
  "Brunei",
  "Bulgaria",
  "Burkina Faso",
  "Burma",
  "Burundi",
  "Cambodia",
  "Cameroon",
  "Canada",
  "Cape Verde",
  "Cayman Islands",
  "Central African Republic",
  "Chad",
  "Chile",
  "China",
  "Christmas Island",
  "Clipperton Island",
  "Cocos (Keeling) Islands",
  "Colombia",
  "Comoros",
  "Congo, Democratic Republic of the",
  "Congo, Republic of the",
  "Cook Islands",
  "Coral Sea Islands",
  "Costa Rica",
  "Cote d'Ivoire",
  "Croatia",
  "Cuba",
  "Cyprus",
  "Czech Republic",
  "Denmark",
  "Dhekelia",
  "Djibouti",
  "Dominica",
  "Dominican Republic",
  "Ecuador",
  "Egypt",
  "El Salvador",
  "Equatorial Guinea",
  "Eritrea",
  "Estonia",
  "Ethiopia",
  "Europa Island",
  "Falkland Islands (Islas Malvinas)",
  "Faroe Islands",
  "Fiji",
  "Finland",
  "France",
  "French Guiana",
  "French Polynesia",
  "French Southern and Antarctic Lands",
  "Gabon",
  "Gambia, The",
  "Gaza Strip",
  "Georgia",
  "Germany",
  "Ghana",
  "Gibraltar",
  "Glorioso Islands",
  "Greece",
  "Greenland",
  "Grenada",
  "Guadeloupe",
  "Guam",
  "Guatemala",
  "Guernsey",
  "Guinea",
  "Guinea-Bissau",
  "Guyana",
  "Haiti",
  "Heard Island and McDonald Islands",
  "Holy See (Vatican City)",
  "Honduras",
  "Hong Kong",
  "Hungary",
  "Iceland",
  "India",
  "Indonesia",
  "Iran",
  "Iraq",
  "Ireland",
  "Isle of Man",
  "Israel",
  "Italy",
  "Jamaica",
  "Jan Mayen",
  "Japan",
  "Jersey",
  "Jordan",
  "Juan de Nova Island",
  "Kazakhstan",
  "Kenya",
  "Kiribati",
  "Korea, North",
  "Korea, South",
  "Kuwait",
  "Kyrgyzstan",
  "Laos",
  "Latvia",
  "Lebanon",
  "Lesotho",
  "Liberia",
  "Libya",
  "Liechtenstein",
  "Lithuania",
  "Luxembourg",
  "Macau",
  "Macedonia",
  "Madagascar",
  "Malawi",
  "Malaysia",
  "Maldives",
  "Mali",
  "Malta",
  "Marshall Islands",
  "Martinique",
  "Mauritania",
  "Mauritius",
  "Mayotte",
  "Mexico",
  "Micronesia, Federated States of",
  "Moldova",
  "Monaco",
  "Mongolia",
  "Montserrat",
  "Morocco",
  "Mozambique",
  "Namibia",
  "Nauru",
  "Navassa Island",
  "Nepal",
  "Netherlands",
  "Netherlands Antilles",
  "New Caledonia",
  "New Zealand",
  "Nicaragua",
  "Niger",
  "Nigeria",
  "Niue",
  "Norfolk Island",
  "Northern Mariana Islands",
  "Norway",
  "Oman",
  "Pakistan",
  "Palau",
  "Panama",
  "Papua New Guinea",
  "Paracel Islands",
  "Paraguay",
  "Peru",
  "Philippines",
  "Pitcairn Islands",
  "Poland",
  "Portugal",
  "Puerto Rico",
  "Qatar",
  "Reunion",
  "Romania",
  "Russia",
  "Rwanda",
  "Saint Helena",
  "Saint Kitts and Nevis",
  "Saint Lucia",
  "Saint Pierre and Miquelon",
  "Saint Vincent and the Grenadines",
  "Samoa",
  "San Marino",
  "Sao Tome and Principe",
  "Saudi Arabia",
  "Senegal",
  "Serbia and Montenegro",
  "Seychelles",
  "Sierra Leone",
  "Singapore",
  "Slovakia",
  "Slovenia",
  "Solomon Islands",
  "Somalia",
  "South Africa",
  "South Georgia and the South Sandwich Islands",
  "Spain",
  "Spratly Islands",
  "Sri Lanka",
  "Sudan",
  "Suriname",
  "Svalbard",
  "Swaziland",
  "Sweden",
  "Switzerland",
  "Syria",
  "Taiwan",
  "Tajikistan",
  "Tanzania",
  "Thailand",
  "Timor-Leste",
  "Togo",
  "Tokelau",
  "Tonga",
  "Trinidad and Tobago",
  "Tromelin Island",
  "Tunisia",
  "Turkey",
  "Turkmenistan",
  "Turks and Caicos Islands",
  "Tuvalu",
  "Uganda",
  "Ukraine",
  "United Arab Emirates",
  "United Kingdom",
  "United States #1",
  "Uruguay",
  "Uzbekistan",
  "Vanuatu",
  "Venezuela",
  "Vietnam",
  "Virgin Islands",
  "Wake Island",
  "Wallis and Futuna",
  "West Bank",
  "Western Sahara",
  "Yemen",
  "Zambia",
  "Zimbabwe",
];

export const COUNTRIES: Array<{
  label: string;
  value: string;
}> = COUNTRY_NAMES.map(name => ({ label: name })).map(suggestion => ({
  label: suggestion.label,
  value: suggestion.label.toUpperCase().replace(" ", "_"),
}));

export const GRAIL =
  "https://grail.com/wp-content/uploads/2016/12/cropped-final-favicon-32x32.png";

export const EXAMPLE_TABLE_DATA = [
  { word: "Argle-bargle", origin: "Britain" },
  { word: "Dwaal", origin: "South Africa" },
  { word: "Floccinaucinihilipilification", origin: "General English" },
  { word: "Lollygag", origin: "North American" },
  { word: "Mouse Potato", origin: "General English" },
  { word: "Ogdoad", origin: "General English" },
  { word: "Sesquipedalian", origin: "General English" },
  { word: "Flibbertigibbet", origin: "Britain" },
  { word: "Curmudgeon", origin: "N/A" },
  { word: "Nudiustertian", origin: "Latin" },
  { word: "Tittynope", origin: "General English" },
  { word: "Pauciloquent", origin: "Britain" },
  {
    word: "Pneumonoultramicroscopicsilicovolcanoconiosis",
    origin: "General English",
  },
  { word: "Bijoux", origin: "Archaic" },
  { word: "Fartlek", origin: "Swedish" },
  { word: "Cruciverbalist", origin: "General English" },
  { word: "Erinaceous", origin: "General English" },
];

export const LARGE_EXAMPLE_TABLE_DATA = [];
for (let i = 0; i < 1000; i++) {
  // Create about 17,000 records.
  LARGE_EXAMPLE_TABLE_DATA.push(...EXAMPLE_TABLE_DATA);
}

export const EXTRA_LARGE_EXAMPLE_TABLE_DATA = [];
for (let i = 0; i < 100; i++) {
  // Create about 1,700,000 records
  EXTRA_LARGE_EXAMPLE_TABLE_DATA.push(...LARGE_EXAMPLE_TABLE_DATA);
}

export const ADAMS_DATA = [
  {
    date: "1848-02-23T16:20:00Z",
    content: <div>John Quincy Adams dies.</div>,
  },
  {
    date: "1818-04-10T12:00:00Z",
    content: (
      <div>
        John Quincy Adams tries to dig to the middle of the Earth to conduct
        trade with the people living there.
      </div>
    ),
  },
  {
    date: "1767-07-11T12:20:00Z",
    content: <div>John Quincy Adams is born.</div>,
  },
];

export const EMU_DATA = [
  {
    date: "1932-12-10T16:20:00Z",
    content: <div>War ends and is declared a victory for the emus.</div>,
  },
  {
    date: "1932-11-08T12:00:00Z",
    content: (
      <div>
        Over 2500 rounds had been fired with the media claiming that only a few
        emus had been killed.
      </div>
    ),
  },
  {
    date: "1932-11-02T12:20:00Z",
    content: (
      <div>
        Sir George Pearce wages war on the local population of emus in
        Australia.
      </div>
    ),
  },
];

export const ALPHABET_ANIMALS_DATA: Array<Array<string | null>> = [
  ["alligator", "bat", "cat", "dog", "elephant", "fox"],
  ["goat", "horse", "iguana", "jaguar", "kangaroo", "lion"],
  ["monkey", "narwhal", "octopus", "penguin", "quail", "raccoon"],
  ["seal", "tiger", "umbrella bird", "vulture", "whale", "x-ray tetra"],
  ["yak", "zebra", null, null, null, null],
];

export const TEST_EXTERNAL_DOMAINS: Map<string, string> = new Map()
  .set("EDC", "edc.com")
  .set("pipeline", "pipeline.com");
