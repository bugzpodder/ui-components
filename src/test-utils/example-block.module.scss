.container {
  margin: 1rem 0;

  .header {
    font-family: monospace;
  }

  .text {
    background: #f6f8fa;
    font-size: 1rem;
    padding: .5rem;
  }
}