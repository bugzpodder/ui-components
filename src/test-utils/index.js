// @flow
export {
  ALPHABET_ANIMALS_DATA,
  ADAMS_DATA,
  COUNTRIES,
  EMU_DATA,
  EXTRA_LARGE_EXAMPLE_TABLE_DATA,
  LARGE_EXAMPLE_TABLE_DATA,
  EXAMPLE_TABLE_DATA,
  GRAIL,
} from "./constants";
export { ExampleBlock } from "./example-block";
export { TestWrapper } from "./test-wrapper";
export { StyleWrapper as ExampleWrapper } from "../style-wrapper";
