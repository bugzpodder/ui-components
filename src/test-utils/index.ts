export {
  ALPHABET_ANIMALS_DATA,
  ADAMS_DATA,
  COUNTRIES,
  COUNTRY_NAMES,
  EMU_DATA,
  EXTRA_LARGE_EXAMPLE_TABLE_DATA,
  LARGE_EXAMPLE_TABLE_DATA,
  EXAMPLE_TABLE_DATA,
  GRAIL,
  TEST_EXTERNAL_DOMAINS,
} from "./constants";
export { ExampleBlock } from "./example-block";
export { TestWrapper } from "./test-wrapper";
export { StyleWrapper as ExampleWrapper } from "../style-wrapper";
