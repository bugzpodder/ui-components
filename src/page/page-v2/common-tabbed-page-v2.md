### Example

```js
import { useState } from "react";
import { ExampleWrapper } from "../../test-utils";
import { ReadOnlyTextField, Alert } from "@grailbio/components";
import { CommonTabbedPageV2 } from "./common-tabbed-page-v2";

const ExampleCommonPage = () => {
  const [activeTab, setActiveTab] = useState("one");
  return (
    <CommonTabbedPageV2
      title="Example Common Page"
      subtitle="This is a subtitle"
      activeTab={activeTab}
      onChangeActiveTab={setActiveTab}
      primaryActions={[
        {
          content: "Button One",
          onClick: () => console.debug("Button One was clicked."),
        },
      ]}
      secondaryActions={[
        {
          content: "Button Two",
          color: "primary",
          onClick: () => console.debug("Button Two was clicked."),
        },
      ]}
      pageConfigs={[
        {
          label: "Tab One",
          key: "one",
          Component: Alert,
          componentProps: {
            message: "Tab One",
            color: "success",
          },
        },
        {
          label: "Tab Two",
          key: "two",
          Component: Alert,
          componentProps: {
            message: "Tab Two",
            color: "info",
          },
        },
      ]}
    >
      <ReadOnlyTextField>Any content can go inside of CommonPage</ReadOnlyTextField>
    </CommonTabbedPageV2>
  );
};

<ExampleWrapper>
  <ExampleCommonPage />
</ExampleWrapper>;
```
