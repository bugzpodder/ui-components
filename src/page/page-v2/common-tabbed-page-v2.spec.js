// @flow
import "jest-dom/extend-expect";
import React, { useState } from "react";
import mockConsole from "jest-mock-console";
import { Alert, CommonSwitch } from "../../index";
import { CommonTabbedPageV2 } from "./common-tabbed-page-v2";
import { TestWrapper } from "../../test-utils";
import { cleanup, fireEvent, render } from "@testing-library/react";

afterEach(cleanup);

const pageConfigs: Array<PageConfig> = [
  {
    label: "Tab One",
    key: "one",
    id: "tab-one",
    Component: Alert,
    componentProps: {
      "data-testid": "alert-one",
      message: "Tab One!",
    },
  },
  {
    label: "Tab Two",
    key: "two",
    id: "tab-two",
    Component: Alert,
    componentProps: {
      "data-testid": "alert-two",
      message: "Tab Two!",
    },
  },
  {
    label: "Tab Three",
    key: "three",
    id: "tab-three",
    Component: Alert,
  },
];

const TestCommonTabbedPage = props => {
  const { mockOnChange, ...otherProps } = props;

  const [activeTab, setActiveTab] = useState("one");
  const setChange = value => {
    mockOnChange(value);
    setActiveTab(value);
  };

  return (
    <TestWrapper>
      <CommonTabbedPageV2
        title="Test Card"
        activeTab={activeTab}
        pageConfigs={pageConfigs}
        primaryActions={[
          {
            content: "Test Action",
          },
        ]}
        secondaryActions={[
          {
            Component: CommonSwitch,
            componentProps: {
              label: "CUSTOM ACTION",
              onChange: () => {},
            },
          },
        ]}
        {...otherProps}
        onChangeActiveTab={setChange}
      >
        <div data-testid="test">Test</div>
      </CommonTabbedPageV2>
    </TestWrapper>
  );
};

test("render common tabbed page", () => {
  const mockOnChange = jest.fn(result => result);
  const { container, queryByTestId, getByTestId } = render(<TestCommonTabbedPage mockOnChange={mockOnChange} />);
  expect(container).toMatchSnapshot();
  expect(getByTestId("alert-one")).toHaveTextContent("Tab One!");
  expect(getByTestId("tab-one")).toHaveAttribute("aria-selected", "true");
  expect(queryByTestId("alert-two")).not.toBeInTheDocument();
  expect(getByTestId("tab-two")).toHaveAttribute("aria-selected", "false");
  expect(getByTestId("spinner-overlay")).toHaveAttribute("data-is-active", "false");
  expect(getByTestId("test")).toBeInTheDocument();

  // Select a new tab
  fireEvent.click(getByTestId("tab-two"));
  expect(mockOnChange).toHaveBeenCalledTimes(1);
  expect(mockOnChange.mock.results[0].value).toEqual("two");
  expect(getByTestId("tab-one")).toHaveAttribute("aria-selected", "false");
  expect(getByTestId("tab-two")).toHaveAttribute("aria-selected", "true");
  expect(queryByTestId("alert-one")).not.toBeInTheDocument();
  expect(queryByTestId("alert-two")).toBeInTheDocument();
  expect(getByTestId("test")).toBeInTheDocument();
  fireEvent.click(getByTestId("tab-three"));
});

test("CommonTabbedPage classes", () => {
  const { getByTestId } = render(<TestCommonTabbedPage classes={{ tabs: "test-tabs", tab: "test-tab" }} />);
  expect(getByTestId("common-tabbed-page-tabs")).toHaveClass("test-tabs");
  expect(getByTestId("tab-one")).toHaveClass("test-tab");
});

test("display loading tabbed page", () => {
  mockConsole();
  const { getByTestId } = render(<TestCommonTabbedPage
    title=""
    isLoading
  />);
  expect(getByTestId("spinner-overlay")).toHaveAttribute("data-is-active", "true");
});

test("require pageConfigs is defined", () => {
  mockConsole();
  expect(() =>
    render(
      <TestWrapper>
        <CommonTabbedPageV2 />
      </TestWrapper>,
    ),
  ).toThrowError();
});

test("activeTab must match at least one page config", () => {
  mockConsole();
  render(<TestCommonTabbedPage activeTab="hi" />);
  // TODO(nsawas): expect error message.
  expect(console.error).toHaveBeenCalled();
});
