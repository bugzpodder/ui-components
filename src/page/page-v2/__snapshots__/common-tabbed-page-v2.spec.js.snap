// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render common tabbed page 1`] = `
<div>
  <div
    class="pageContainer"
    data-testid="common-page"
  >
    <div
      class="MuiPaper-root stickyHeader pageHeader MuiPaper-elevation1"
      data-testid="common-page-header"
    >
      <div
        class="titleContainer"
        data-testid="common-page-title-container"
      >
        <h6
          class="MuiTypography-root title MuiTypography-subtitle1 MuiTypography-noWrap"
          data-testid="common-page-title"
        >
          Test Card
        </h6>
        
      </div>
      <div
        class="centerHeader"
      >
        <div
          class="MuiTabs-root tabsContainer"
          data-testid="common-tabbed-page-tabs"
        >
          <div
            class="PrivateTabScrollButton-root-69 MuiTabs-scrollButtons scrollButton"
          />
          <div
            class="MuiTabs-scrollable"
            style="width: 99px; height: 99px; position: absolute; top: -9999px; overflow: scroll;"
          />
          <div
            class="MuiTabs-scroller MuiTabs-scrollable"
            style="margin-bottom: 0px;"
          >
            <div
              class="MuiTabs-flexContainer tabs"
              role="tablist"
            >
              <button
                aria-selected="true"
                class="MuiButtonBase-root MuiTab-root tab tabbed-page-tab MuiTab-textColorPrimary Mui-selected"
                data-testid="tab-one"
                id="tab-one"
                role="tab"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiTab-wrapper"
                >
                  Tab One
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root tab tabbed-page-tab MuiTab-textColorPrimary"
                data-testid="tab-two"
                id="tab-two"
                role="tab"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiTab-wrapper"
                >
                  Tab Two
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root tab tabbed-page-tab MuiTab-textColorPrimary"
                data-testid="tab-three"
                id="tab-three"
                role="tab"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiTab-wrapper"
                >
                  Tab Three
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <span
              class="PrivateTabIndicator-root-84 PrivateTabIndicator-colorPrimary-85 MuiTabs-indicator"
              style="left: 0px; width: 0px;"
            />
          </div>
          <div
            class="PrivateTabScrollButton-root-69 MuiTabs-scrollButtons scrollButton"
          />
        </div>
      </div>
      <div
        class="headerActions"
        data-testid="common-page-header-actions"
      >
        <div
          class="primaryActions"
          data-testid="common-page-primary-actions"
        >
          <button
            class="MuiButtonBase-root MuiButton-root primaryAction MuiButton-contained MuiButton-containedPrimary"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Test Action
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <button
          aria-describedby="test"
          class="MuiButtonBase-root MuiIconButton-root secondaryActionsButton"
          data-testid="common-page-secondary-actions-button"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="content"
      data-testid="common-page-content"
    >
      <div
        class="alertContainer"
        data-testid="alert"
      >
        <div
          class="MuiTypography-root MuiPaper-root MuiPaper-elevation6 MuiSnackbarContent-root error alert-error MuiTypography-body2"
          data-testid="alert-one"
          role="alertdialog"
        >
          <div
            class="MuiSnackbarContent-message"
          >
            <span
              data-testid="error-alert-message"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root alertIcon"
                data-testid="error-alert-icon"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
                />
              </svg>
              Tab One!
            </span>
          </div>
        </div>
      </div>
      <div
        data-testid="test"
      >
        Test
      </div>
      <div
        class="spinnerOverlay"
        data-is-active="false"
        data-testid="spinner-overlay"
      >
        <div
          class="loading"
        />
      </div>
    </div>
  </div>
</div>
`;
